lib/transfer.c

```diff
@@ -918,13 +918,42 @@ static CURLcode readwrite_upload(struct Curl_easy *data,
 
   do {
     curl_off_t nbody;
+    ssize_t offset = 0;
+
+    if(0 != k->upload_present && k->upload_present < 1024 && !k->upload_done) {
+      DEBUGASSERT(data->set.upload_buffer_size >= 1024);
+      offset = k->upload_present;
+#ifndef CURL_DISABLE_HTTP
+      /* disable appending to upload buffer if HTTP body not ready */
+      if(conn->handler->protocol & PROTO_FAMILY_HTTP) {
+        if(conn->bits.authneg)
+          offset = 0;
+#ifdef USE_HYPER
+        offset = 0;
+#else
+        if(data->state.trailers_state != TRAILERS_NONE ||
+           k->p.http->sending != HTTPSEND_BODY ||
+           k->upload_chunky)
+          offset = 0;
+        else if(data->state.httpreq == HTTPREQ_POST ||
+                data->state.httpreq == HTTPREQ_POST_FORM ||
+                data->state.httpreq == HTTPREQ_POST_MIME) {
+          if(!k->p.http->postdata)
+            offset = 0;
+        }
```

Comment:
Can you elaborate on why these condition should use a zero offset?

Reply:
`offset = 0` indicates "do not append": preserve existing behavior and falling through to send whatever remains in the buffer until the buffer is empty, and then read more.

Reply:
`k->p.http->postdata` is NULL in the crash in test363.  I do not understand why that is the case when `k->p.http->sending == HTTPSEND_BODY`

The other conditions are to avoid situations where it might not be safe to append.  For example, the current chunk code writes a variable amount for the chunked header, and it adjusts k->upload_fromhere.  Were we to modify the chunked code to write a fixed length header, then appending might be possible.  I was not sure about the safety of appending with the trailer state, and it seemed prudent to me to restrict appending for HTTP to when sending `k->p.http->sending == HTTPSEND_BODY`.  I am happy to be corrected on any of this.

Reply:
ok, I figured out that the reason test363 was crashing was that the POST body had already been queued, and k->upload_done *would be* set after the existing k->upload_present was sent.  I have adjusted the conditions and forced-pushed an update to this PR.

If I adjust the code writing chunked header to use a fixed size, I could make this patch work when k->upload_chunky is set.  Would you like me to add a patch to do that?  Using a fixed size might send up to 7 excess bytes in the chunked header for each chunk sent.

Reply:
For your review, I pushed an additional commit which uses a fixed-size chunked header, so that appending to the buffer is safe for HTTP/1.1 chunked, too.  I am not sure whether or not the additional bytes in the chunked header are a worthwhile tradeoff, and would lean toward *not including* this additional commit, but I'll leave that up to you.